processing/src/main/java/org/apache/druid/segment/UnnestColumnValueSelectorCursor.java

```diff
@@ -0,0 +1,334 @@
+          return baseColumSelectorFactory.makeDimensionSelector(dimensionSpec);
+        }
+        throw new UOE("Unsupported dimension selector while using column value selector for column [%s]", outputName);
+      }
+
+      @Override
+      public ColumnValueSelector makeColumnValueSelector(String columnName)
+      {
+        if (!outputName.equals(columnName)) {
+          return baseColumSelectorFactory.makeColumnValueSelector(columnName);
+        }
+        return new ColumnValueSelector()
+        {
+          @Override
+          public double getDouble()
+          {
+            Object value = getObject();
+            if (value == null) {
+              return 0;
+            }
+            if (value instanceof Number) {
+              return ((Number) value).doubleValue();
+            }
+            throw new UOE("Cannot convert object to double");
+          }
+
+          @Override
+          public float getFloat()
+          {
+            Object value = getObject();
+            if (value == null) {
+              return 0;
+            }
+            if (value instanceof Number) {
+              return ((Number) value).floatValue();
+            }
+            throw new UOE("Cannot convert object to float");
+          }
+
+          @Override
+          public long getLong()
+          {
+            Object value = getObject();
+            if (value == null) {
+              return 0;
+            }
+            if (value instanceof Number) {
+              return ((Number) value).longValue();
+            }
+            throw new UOE("Cannot convert object to long");
+          }
+
+          @Override
+          public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+          {
+            columnValueSelector.inspectRuntimeShape(inspector);
+          }
+
+          @Override
+          public boolean isNull()
+          {
+            return getObject() == null;
+          }
+
+          @Nullable
+          @Override
+          public Object getObject()
+          {
+            if (!unnestListForCurrentRow.isEmpty()) {
+              if (allowSet == null || allowSet.isEmpty()) {
+                return unnestListForCurrentRow.get(index);
+              } else if (allowSet.contains((String) unnestListForCurrentRow.get(index))) {
+                return unnestListForCurrentRow.get(index);
+              }
+            }
+            return null;
+          }
+
+          @Override
+          public Class<?> classOfObject()
+          {
+            return Object.class;
+          }
+        };
+      }
+
+      @Nullable
+      @Override
+      public ColumnCapabilities getColumnCapabilities(String column)
+      {
+        if (!outputName.equals(column)) {
+          return baseColumSelectorFactory.getColumnCapabilities(column);
+        }
+        // This currently returns the same type as of the column to be unnested
+        // This is fine for STRING types
```

Comment:
nit: I think its not really great for string types either since if 'hasMultipleValues` is set the engine will take less efficient paths and treat it as a multi-value string.

Btw, this is pretty easy to fix, all you need to do is something like this, but is fine to do as a follow-up too
```java
      final ColumnCapabilities capabilities = baseColumSelectorFactory.getColumnCapabilities(columnName);
      if (capabilities.isArray()) {
        return ColumnCapabilitiesImpl.copyOf(capabilities).setType(capabilities.getElementType());
      }
      if (capabilities.hasMultipleValues()) {
        return ColumnCapabilitiesImpl.copyOf(capabilities).setHasMultipleValues(false);
      }
      return capabilities;
```

Reply:
Thanks I'll do this in the followup PR. Will work on the runner tests too.